### PR TITLE
enhancement(transforms): New filter `datadog_search` condition type

### PR DESCRIPTION
**Pull request by leebenson**
Follow-on to #7633 (currently targeting the open PR branch.)

Adds a new `condition.type = "datadog_search"` option to the `filter `transform, for parsing [Datadog Search Syntax](https://docs.datadoghq.com/logs/search_syntax/).

## Example config

```toml
[api]
  enabled = true

[sources.gen1]
  type = "generator"
  format = "json"
  batch_interval = 0.2

[transforms.filter]
  type = "filter"
  inputs = ["gen1"]
  condition.type = "datadog_search"
  condition.source = "*mopper"

[sinks.blackhole]
  type = "blackhole"
  inputs = ["gen1"]
  print_amount = 1000
```

Running `vector tap filter` locally returns lines matching a _mopper_ suffix:

```
{"message":"{\"host\":\"11.218.4.226\",\"user-identifier\":\"Karimmove\",\"datetime\":\"11/Jun/2021:15:08:31\",\"method\":\"OPTION\",\"request\":\"/booper/bopper/mooper/mopper\",\"protocol\":\"HTTP/2.0\",\"status\":\"301\",\"bytes\":12808,\"referer\":\"https://random.de/secret-info/open-sesame\"}","timestamp":"2021-06-11T14:08:31.802773Z"}
{"message":"{\"host\":\"134.56.164.243\",\"user-identifier\":\"shaneIxD\",\"datetime\":\"11/Jun/2021:15:08:32\",\"method\":\"HEAD\",\"request\":\"/booper/bopper/mooper/mopper\",\"protocol\":\"HTTP/1.0\",\"status\":\"304\",\"bytes\":21192,\"referer\":\"https://up.us/apps/deploy\"}","timestamp":"2021-06-11T14:08:32.205288Z"}
{"message":"{\"host\":\"80.162.123.16\",\"user-identifier\":\"devankoshal\",\"datetime\":\"11/Jun/2021:15:08:32\",\"method\":\"GET\",\"request\":\"/wp-admin\",\"protocol\":\"HTTP/2.0\",\"status\":\"503\",\"bytes\":24379,\"referer\":\"https://names.net/booper/bopper/mooper/mopper\"}","timestamp":"2021-06-11T14:08:32.803780Z"}
{"message":"{\"host\":\"137.139.222.233\",\"user-identifier\":\"Karimmove\",\"datetime\":\"11/Jun/2021:15:08:33\",\"method\":\"HEAD\",\"request\":\"/booper/bopper/mooper/mopper\",\"protocol\":\"HTTP/2.0\",\"status\":\"400\",\"bytes\":41396,\"referer\":\"https://make.com/controller/setup\"}","timestamp":"2021-06-11T14:08:33.005674Z"}
```

Marking this as draft to add fuller coverage against the [unit tests](https://github.com/timberio/vector/blob/31df455187a6de844579f78a82fc9afc143f033d/lib/datadog/search-syntax/src/builder.rs#L213-L475) in #7633.

### CODE REVIEW

**Comment by binarylogic** [comment]
Nice!

> condition.type = "datadog"

Would it be more clear to rename this to `datadog_search`?

**Comment by leebenson** [comment]
> Would it be more clear to rename this to `datadog_search`?

👍🏻  It's a little ambiguous at present. I've referred to it as:

- Datadog syntax
- Log syntax
- Search syntax

And a combination of the above, conversationally. Wasn't sure if there was a de facto/preferred label.

**Comment by binarylogic** [comment]
I'd prefer `datadog_search` since it makes more sense in the broader context of Vector and that's the terminology Datadog uses.

**Comment by leebenson** [comment]
Marking this as ready to review.

There's still a few unit tests against more complex queries I'd like to add, but the transform logic itself should be good to look at.

File: src/conditions/datadog_search.rs [hunk 1]
```diff
@@ -0,0 +1,658 @@
...
+                log_event!["custom" => r#"{"title": "foobla"}"#],
+            ),
+            // Multiple wildcards (negate w/-).
+            (
+                "-*b*la*",
+                log_event![],
+                log_event!["custom" => r#"{"title": "foobla"}"#],
+            ),
+            // Wildcard prefix - tag.
+            (
```
Comment (vladimir-dd) [inline src/conditions/datadog_search.rs:357]: I guess we need to updates these tests after the most recent search syntax changes, right?

Comment (vladimir-dd) [reply src/conditions/datadog_search.rs:357]: The match use-case should be something like: `log_event!["tags" => vec!["a:foobla"]]`

File: src/conditions/datadog_search.rs [hunk 2]
```diff
@@ -0,0 +1,658 @@
...
+            ),
+            // Range - numeric, inclusive, facet (negate).
+            (
+                "NOT @b:[1 TO 10]",
+                log_event!["custom" => json!({"b": 11})],
+                log_event!["custom" => json!({"b": 5})],
+            ),
+            // Range - numeric, inclusive, facet (negate w/-).
+            (
+                "-@b:[1 TO 10]",
```
Comment (vladimir-dd) [inline src/conditions/datadog_search.rs:628]: As we discussed offline, here we should also make sure that for string values we do alphanumerical comparisons.

**Comment by vladimir-dd** [comment]
Overall looks great 👍

File: lib/datadog/search-syntax/src/field.rs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        return DEFAULT_FIELDS
+            .iter()
+            .map(|s| Field::Default((*s).to_owned()))
+            .collect();
+    }
+
+    let field = match value.replace("@", "custom.") {
+        v if DEFAULT_FIELDS.contains(&v.as_ref()) => Field::Default(v),
+        v if RESERVED_ATTRIBUTES.contains(&v.as_ref()) => Field::Reserved(v),
+        v if value.starts_with('@') => Field::Facet(v),
```
Comment (vladimir-dd) [inline lib/datadog/search-syntax/src/field.rs:67]: does it make sense to move this `starts_with` check up for performance reasons?

File: src/conditions/datadog_search.rs [hunk 4]
```diff
@@ -0,0 +1,749 @@
...
+            ),
+            // Range - alpha, inclusive, facet.
+            (
+                "@b:[a TO z]",
+                log_event!["custom" => json!({"b": "c"})],
+                log_event!["custom" => json!({"b": 5})],
+            ),
+            // Range - alphanumeric, inclusive, facet.
+            (
+                r#"@b:["1" TO "100"]"#,
```
Comment (vladimir-dd) [inline src/conditions/datadog_search.rs:641]: Here we should also support `@b:[1 TO 100]` - the same logic as `@b:["1" TO "100"]` provided `b` is a string value. (Sorry, I should've emphasised this difference better).

Comment (leebenson) [reply src/conditions/datadog_search.rs:641]: That should actually be the case already 👍🏻

Comment (leebenson) [reply src/conditions/datadog_search.rs:641]: To be clear:

```
@b:[1 TO 10]
```

Is translated to:

```
(((((is_integer(.custom.b) || is_float(.custom.b)) && .custom.b >= 1) || .custom.b >= "1") && (((is_integer(.custom.b) || is_float(.custom.b)) && .custom.b <= 10) || .custom.b <= "10")) ?? false)
```

Is that right?

Comment (vladimir-dd) [reply src/conditions/datadog_search.rs:641]: can we add a test case with `@b:[1 TO 100]` to make sure it works as expected?

**Review by vladimir-dd — APPROVED** [review]
Nice work 👍

**Pull request merged**